### DISCUSSION TITLE
image url retrival process in strapi is confusing

### DISCUSSION THREAD

**Discussion by cadentic** [vercel/next.js#40486] [Help]
hi

i literally wonder how we can access image url from the strapi ? i tried to open a thread in https://forum.strapi.io/t/how-to-retrive-the-image-url-typeerror-cannot-read-properties-of-undefined-reading-formats/21773/7 but it is not retrieving any images. 

``` <CardMedia component="img" height="200" image="{{pages.attributes.images.url}}" alt="Paella dish" /> ```

![Capture_struct](https://user-images.githubusercontent.com/55697529/189719545-7a39f2bd-3136-47da-bdec-fccbd9b2e9f4.JPG)
[testmonial componant](https://github.com/cadentic/strapi-webdev/blob/main/zettadaten_fe_next17/components/Testimonial.js)

**Comment by wwwdepot** [comment]
Strapi REST API does not include Images or Relation fields by default, you need add a `populate` parameter to the request to get them.

```js
const API_URL = 'http://localhost:1337/api/pages?populate=*'; // populate all Relation & Media fields
const res = await axios.get(API_URL);
```

Read https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest/populating-fields.html#relation-media-fields

**Reply by cadentic** [reply]
this shows an error while returning them `
```
Home.getInitialProps = async ctx => {
  try {
    {/*
    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;
//    console.log(Pages);     
    return { Pages};
  //  return { HomeAccordition };
  } catch (error) {
    return { error };
  }
*/}
    let Pages =
      "'http://localhost:1337/api/' + 'pages'";
   // let homeAaccorditions =
   //   "'http://localhost:1337/api/' + 'home-accorditions'";
   // let Menus =
    //  "'http://localhost:1337/api/' + 'menus'";

   // let HeroSections =
   //   "'http://localhost:1337/api/' + 'hero-sections'";
    let myimages = "'http://localhost:1337/api/pages?populate=*'";
    const requestPages = axios.get(Pages);
   // const requesthomeAaccorditions = axios.get(homeAaccorditions);
  //  const requestMenus = axios.get(Menus);
  //  const requestHeroSections = axios.get(HeroSections);
    const requestMyImages = axios.get(myimages);
    axios
      .all([requestPages, requestMyImages ]) // requesthomeAaccorditions, requestMenus, requestHeroSections])
      .then(
        axios.spread((...responses) => {
          const requestPages = responses[0];
          //  const requesthomeAaccorditions = responses[1];
          //  const requestMenus = responses[2];
          //  const requestHeroSections = responses[3];
          requestMyImages = responses[2];

          // use/access the results
      
          // console.log(requestPages, requesthomeAaccorditions, requestMenus, requestHeroSections);
          return { requestPages, requestMyImages }; //requesthomeAaccorditions, requestMenus, requestHeroSections };
        })
     
  )
  } catch (error) {
    return { error };
  }

};
```
![Capture_undefined_object](https://user-images.githubusercontent.com/55697529/191272022-ca196d48-3e0c-49d1-b24b-72a255fe13b6.JPG)

`

**Reply by wwwdepot** [reply]
Yea, you didn't return anything from the getInitialProps function, that's why it's throwing error. Try a simple page

```jsx
import axios from "axios";

export default function StrapiPage(props) {
  return <pre>{JSON.stringify(props, null, 4)}</pre>;
}

StrapiPage.getInitialProps = async () => {
  try {
    let API_ENDPOINT = "http://localhost:1337/api/pages?populate=*";
    const pageData = await axios.get(API_ENDPOINT).then((res) => res.data);

    return {
      props: {
        pageData,
      },
    };
  } catch (error) {
    throw error;
  }
};
```

**Reply by cadentic** [reply]
if i do so i will not have proper access to the ```Pages.attribute.title``` components so. i went for othe other version and retuned them as per es6 syntax with the array.

**Reply by cadentic** [reply]
@wwwdepot 

```Home.getInitialProps = async ctx => {
  try {

    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;
    const res2 = await axios.get('http://localhost:1337/api/pages?populate=*')
    //    console.log(Pages);
    const images = res2.data.;
    return { Pages, images };
    //  return { HomeAccordition };
  } catch (error) {
    return { error };
  }
```
even this one is not working for images. 

bcz ```{requestMyImages.attributes.image.url}``` is not populating within the componant. 
[if you want to correct me you may look at this code once ](https://github.com/cadentic/strapi-webdev/blob/main/zettadaten_fe_next17/pages/index.js)

**Reply by cadentic** [reply]
> Yea, you didn't return anything from the getInitialProps function, that's why it's throwing error. Try a simple page
> 
> ```js
> import axios from "axios";
> 
> export default function StrapiPage(props) {
>   return <pre>{JSON.stringify(props, null, 4)}</pre>;
> }
> 
> StrapiPage.getInitialProps = async () => {
>   try {
>     let API_ENDPOINT = "http://localhost:1337/api/pages?populate=*";
>     const pageData = await axios.get(API_ENDPOINT).then((res) => res.data);
> 
>     return {
>       props: {
>         pageData,
>       },
>     };
>   } catch (error) {
>     throw error;
>   }
> };
> ```

this thing works though as expected but I need the array traversal process to access the images within components.

**Reply by cadentic** [reply]
@wwwdepot interestingly simple thing works always but keeping things simple is not that simple. simplest things are not welcomed by recruiters as they think the simplest task recruiters will do it on their own so techies are need to look for complex task eventually.
